### ISSUE TITLE
Test fails with Wagtail 2beta 

### ISSUE THREAD

**Issue by jossingram** [wagtail/wagtail#4232]
### Issue Summary

A "self.assertCanCreate(self.root, ArticlePage, {" line which has this below to test streamfield content is failing on the google_map line.

     'page_content': json.dumps([
                    {"type": "heading", "value": "Heading Test"},
                    {"type": "embedded_video", "value": "https://xxxxxxxxxxxxxxxxxxx"},
                    {"type": "fox_stories", "value": {"display_mode": "standard", "tagged_by_keyword": "business", "stories_limit": "4"}},
                    {"type": "paragraph", "value": "<p>This is text</p>"},
                    {"type": "google_map", "value": {"map_pins": [], "map_lat": "12.3434", "map_zoom_level": "14", "map_long": "-52.34334"}}
                ]),

Error from running tests: 

    Traceback (most recent call last):
      File "/code/researchsite/tests/test_create_pages.py", line 64, in test_can_create_article_page
        {"type": "google_map", "value": {"map_pins": [], "map_lat": "12.3434", "map_zoom_level": "14", "map_long": "-52.34334"}}
      File "/usr/local/lib/python3.5/dist-packages/wagtail/tests/utils.py", line 275, in assertCanCreate
        response = self.client.post(add_url, data, follow=True)
      File "/usr/local/lib/python3.5/dist-packages/django/test/client.py", line 548, in post
        secure=secure, **extra)
      File "/usr/local/lib/python3.5/dist-packages/django/test/client.py", line 350, in post
        secure=secure, **extra)
      File "/usr/local/lib/python3.5/dist-packages/django/test/client.py", line 416, in generic
        return self.request(**r)
      File "/usr/local/lib/python3.5/dist-packages/django/test/client.py", line 501, in request
        six.reraise(*exc_info)
      File "/usr/local/lib/python3.5/dist-packages/django/utils/six.py", line 686, in reraise
        raise value
      File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/exception.py", line 41, in inner
        response = get_response(request)
      File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
        response = self._get_response(request)
      File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py", line 187, in _get_response
        response = self.process_exception_by_middleware(e, request)
      File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py", line 185, in _get_response
        response = wrapped_callback(request, *callback_args, **callback_kwargs)
      File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/cache.py", line 43, in _cache_controlled
        response = viewfunc(request, *args, **kw)
      File "/usr/local/lib/python3.5/dist-packages/wagtail/admin/urls/__init__.py", line 95, in wrapper
        return view_func(request, *args, **kwargs)
      File "/usr/local/lib/python3.5/dist-packages/wagtail/admin/decorators.py", line 28, in decorated_view
        return view_func(request, *args, **kwargs)
      File "/usr/local/lib/python3.5/dist-packages/wagtail/admin/views/pages.py", line 194, in create
        if form.is_valid():
      File "/usr/local/lib/python3.5/dist-packages/modelcluster/forms.py", line 308, in is_valid
        form_is_valid = super(ClusterForm, self).is_valid()
      File "/usr/local/lib/python3.5/dist-packages/django/forms/forms.py", line 183, in is_valid
        return self.is_bound and not self.errors
      File "/usr/local/lib/python3.5/dist-packages/django/forms/forms.py", line 175, in errors
        self.full_clean()
      File "/usr/local/lib/python3.5/dist-packages/django/forms/forms.py", line 384, in full_clean
        self._clean_fields()
      File "/usr/local/lib/python3.5/dist-packages/django/forms/forms.py", line 396, in _clean_fields
        value = field.widget.value_from_datadict(self.data, self.files, self.add_prefix(name))
      File "/usr/local/lib/python3.5/dist-packages/wagtail/core/blocks/base.py", line 512, in value_from_datadict
        return self.block_def.value_from_datadict(data, files, name)
      File "/usr/local/lib/python3.5/dist-packages/wagtail/core/blocks/stream_block.py", line 162, in value_from_datadict
        child_block.value_from_datadict(data, files, '%s-%d-value' % (prefix, i)),
      File "/usr/local/lib/python3.5/dist-packages/wagtail/core/blocks/field_block.py", line 69, in value_from_datadict
        return self.value_from_form(self.field.widget.value_from_datadict(data, files, prefix))
      File "/usr/local/lib/python3.5/dist-packages/wagtail/admin/rich_text/editors/draftail/__init__.py", line 56, in value_from_datadict
        return self.converter.to_database_format(original_value)
      File "/usr/local/lib/python3.5/dist-packages/wagtail/admin/rich_text/converters/contentstate.py", line 93, in to_database_format
        return self.exporter.render(json.loads(contentstate_json))
      File "/usr/lib/python3.5/json/__init__.py", line 319, in loads
        return _default_decoder.decode(s)
      File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
        obj, end = self.raw_decode(s, idx=_w(s, 0).end())
      File "/usr/lib/python3.5/json/decoder.py", line 357, in raw_decode
        raise JSONDecodeError("Expecting value", s, err.value) from None
    json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

Not sure what this is, possibly not a bug but something I've done wrong, think I've renamed all the module names though as in the release notes. 

### Technical details

* Python version: Python3.5
* Django version:1.11.9
* Wagtail version: 2 beta

**Comment by gasman** [comment]
Thanks for the report @jossingram. It looks like this is failing on the rich text field - presumably because the new Draftail editor works with a JSON representation of the text, and so the code to support it isn't expecting to receive the HTML string (`<p>This is text</p>`).

Unfortunately the JSON representation isn't particularly human-readable/writable, which goes against the principle of assertCanCreate (which expects you to pass the page content in the same format that the browser would submit it). I'll have to ponder the best way to resolve this...

**Comment by gasman** [comment]
@jossingram Would you be willing to share the complete `self.assertCanCreate` statement from your test (with data anonymised if necessary), please? That would give me a better idea of how `assertCanCreate` is being used "in the wild", and the best way to update its behaviour to fix this.

I suspect the fix will probably involve adding 'translation' wrapper functions around rich text content, to convert the human-readable HTML-ish representation into the contentState JSON used by the editor internally, something like:

    'page_content': json.dumps([
        # ...
        {"type": "paragraph", "value": html_to_contentstate("<p>This is text</p>")},
    ]),

As this is a breaking change, we'll provide details of how to update existing tests as part of the 2.0 release notes.

@timheap - just wondering if you have any thoughts on other approaches we can take here (as the original developer of `assertCanCreate` et al)? Stated more generally, the problem here is that `assertCanCreate` ultimately relies on the developer hand-crafting a form POST body, which is an interface that wasn't built with user-friendliness or API stability in mind. In this case, changing the representation of rich text fields is what's broken it, but I dare say there are other places where the POST representation is rather intractable without knowledge of Wagtail / Django internals (inline panels / formset management forms is another that springs to mind). One possibility is that we bypass the admin view entirely, and have `assertCanCreate` test creation at the model level instead - but I guess that undermines most or all of the reason for `assertCanCreate` to exist in the first place...

**Comment by timheap** [comment]
@gasman Your suggestion sounds sensible. In the past, I've created similar helpers to make the creation of inline panel and stream field data easier. I've put the code for the helpers [in this gist](https://gist.github.com/timheap/02cb81557526fbbf49e9c90583bbabac). If you're interested in putting something like this in Wagtail, feel free to reuse the code. Adding another helper to turn HTML in to Draft.js JSON would be a sensible addition to these helpers.

The helpers from the gist can be used as follows, a test case copied from an actual project:

```python
from project.models import ContentPage
from project.tests.base import BasicSiteTestCase
from project.tests.utils import nested_form_data, streamfield


class TestContentPageAdmin(BasicSiteTestCase):
    def test_create_page(self):
        self.assertCanCreate(self.home_page, ContentPage, nested_form_data({
            'title': 'Content Page',
            'banner_image': self.banner_image.pk,
            'body': streamfield([
                ('text', '<p>hello, world!</p>'),
                ('heading1', 'Heading'),
                ('block_quote', {
                    'quote': 'Statement',
                    'attribution': 'Person',
                }),
                ('heading2', 'Heading'),
                ('image_with_caption', {
                    'image': self.banner_image.pk,
                    'caption': 'Caption',
                }),
                ('embed', 'https://www.youtube.com/watch?v=aNW4vDgQW7M'),
            ]),
            'related_things': inline_formset([
                {'thing': 'foo'},
                {'thing': 'bar'},
            ]),
        }))

        # Make sure it renders
        page = ContentPage.objects.get()
        response = self.client.get(page.url)
        self.assertEqual(response.status_code, 200)
```

**Comment by jossingram** [comment]
Here is the full use

```
class MyPageTests(WagtailPageTests):
    def setUp(self):
        super(MyPageTests, self).setUp()
        site = Site.objects.get(is_default_site=True)
        self.root = site.root_page.specific

    def test_can_create_article_page(self):
        self.assertCanCreate(self.root, ArticlePage, {
            'title': 'About us',
            'side_nav':True,
            'side_nav_mode':'siblings',
            'page_content-count': 5,
            'page_content-0-deleted': '',
            'page_content-0-order': 0,
            'page_content-0-type': 'heading',
            'page_content-0-value': 'Heading Test',
            'page_content-1-deleted':'',
            'page_content-1-order':1,
            'page_content-1-type':'embedded_video',
            'page_content-1-value': 'https://www.youtube.com/watch?v=xxxxxxxx',
            'page_content-2-deleted':'',
            'page_content-2-order':2,
            'page_content-2-type':'fox_stories',
            'page_content-2-value-tagged_by_keyword':'business',
            'page_content-2-value-display_mode':'standard',
            'page_content-2-value-stories_limit':'4',
            'page_content-3-deleted':'',
            'page_content-3-order':3,
            'page_content-3-type':'paragraph',
            'page_content-3-value': '<p>This is text</p>',
            'page_content-4-deleted':'',
            'page_content-4-order':4,
            'page_content-4-type':'google_map',
            'page_content-4-value-map_pins':'',
            'page_content-4-value-map_lat':'12.3434',
            'page_content-4-value-map_zoom_level':'14',
            'page_content-4-value-map_long':'-52.34334',
            'page_content-4-value-map_pins-count':0,
            'related_links-TOTAL_FORMS': 0,
            'related_links-INITIAL_FORMS': 0,
            'related_links-MIN_NUM_FORMS': 0,
            'related_links-MAX_NUM_FORMS': 1000,
            'page_content': json.dumps([
                {"type": "heading", "value": "Heading Test"},
                {"type": "embedded_video", "value": "https://www.youtube.com/watch?v=xxxxxxxx"},
                {"type": "fox_stories", "value": {"display_mode": "standard", "tagged_by_keyword": "business", "stories_limit": "4"}},
                {"type": "paragraph", "value": "<p>This is text</p>"},
                {"type": "google_map", "value": {"map_pins": [], "map_lat": "12.3434", "map_zoom_level": "14", "map_long": "-52.34334"}}
            ]),
        })
```

**Comment by gasman** [comment]
PR now created: #4287

@jossingram With the new helper functions introduced here, your test will now become:

    from wagtail.tests.utils.form_data import inline_formset, nested_form_data, rich_text, streamfield

    # ...

    def test_can_create_article_page(self):
        self.assertCanCreate(self.root, ArticlePage, nested_form_data({
            'title': 'About us',
            'side_nav':True,
            'side_nav_mode':'siblings',
            'page_content': streamfield([
                ('heading', 'Heading Test'),
                ('embedded_video', 'https://www.youtube.com/watch?v=xxxxxxxx'),
                ('fox_stories', {'tagged_by_keyword': 'business', 'display_mode': 'standard', 'stories_limit': '4'}),
                ('paragraph', rich_text('<p>This is text</p>')),
                ('google_map', {'map_pins': '', 'map_lat': '12.3434', 'map_zoom_level': '14', 'map_long': '-52.34334', 'map_pins-count': 0})
            ]),
            'related_links': inline_formset([]),
        }))

(I'm not 100% sure I've got the StructBlocks right - will have to add another test case for those. Having said that, wrapping `'<p>This is text</p>'` as `rich_text('<p>This is text</p>')` should be the only change required to make the test pass.)

Incidentally, I'm pretty sure the `page_content` JSON is redundant in your code above - the form processing logic will pick up the values from the individual fields instead.

**Comment by jossingram** [comment]
Thanks, I'll amend my tests

**Comment by gasman** [comment]
Fixed in #4287
